Point at argument instead of call for their obligations

estebank opened the pull request:
When an obligation is introduced by a specific `fn` argument, point at
the argument instead of the `fn` call if the obligation fails to be
fulfilled.

Move the information about pointing at the call argument expression in
an unmet obligation span from the `FulfillmentError` to a new
`ObligationCauseCode`.

When giving an error about an obligation introduced by a function call
that an argument doesn't fulfill, and that argument is a block, add a
span_label pointing at the innermost tail expression.

Current output:

```
error[E0425]: cannot find value `x` in this scope
 --> f10.rs:4:14
  |
4 |         Some(x * 2)
  |              ^ not found in this scope

error[E0277]: expected a `FnOnce<({integer},)>` closure, found `Option<_>`
 --> f10.rs:2:31
  |
2 |       let p = Some(45).and_then({
  |  ______________________--------_^
  | |                      |
  | |                      required by a bound introduced by this call
3 | |         |x| println!("doubling {}", x);
4 | |         Some(x * 2)
  | |         -----------
5 | |     });
  | |_____^ expected an `FnOnce<({integer},)>` closure, found `Option<_>`
  |
  = help: the trait `FnOnce<({integer},)>` is not implemented for `Option<_>`
```

Previous output:

```
error[E0425]: cannot find value `x` in this scope
 --> f10.rs:4:14
  |
4 |         Some(x * 2)
  |              ^ not found in this scope

error[E0277]: expected a `FnOnce<({integer},)>` closure, found `Option<_>`
 --> f10.rs:2:22
  |
2 |     let p = Some(45).and_then({
  |                      ^^^^^^^^ expected an `FnOnce<({integer},)>` closure, found `Option<_>`
  |
  = help: the trait `FnOnce<({integer},)>` is not implemented for `Option<_>`
```

Partially address #27300. Will require rebasing on top of #88546.

estebank committed:
Point at argument instead of call for their obligations

When an obligation is introduced by a specific `fn` argument, point at
the argument instead of the `fn` call if the obligation fails to be
fulfilled.

estebank committed:
Refactor `FulfillmentError` to track less data

Move the information about pointing at the call argument expression in
an unmet obligation span from the `FulfillmentError` to a new
`ObligationCauseCode`.

estebank committed:
Point at call span that introduced obligation for the arg

estebank committed:
Account for blocks in arguments

When giving an error about an obligation introduced by a function call
that an argument doesn't fulfill, and that argument is a block, add a
span_label pointing at the innermost tail expression.

rust-highfive commented:
r? @nagisa

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3533406343) [(plain)](https://github.com/rust-lang/rust/commit/00757fc819f7f3f494689e7fb58a8af021fe26ef/checks/3533406343/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
error[E0308]: mismatched types
   --> compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:494:32
    |
494 |             std::rc::Rc::clone(parent_code)
    |
    = note: expected reference `&Rc<_>`
    = note: expected reference `&Rc<_>`
               found reference `&Arc<rustc_middle::traits::ObligationCauseCode<'_>>`
error[E0308]: mismatched types
   --> compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:683:32
    |
    |
683 |             std::rc::Rc::clone(parent_code)
    |
    = note: expected reference `&Rc<_>`
    = note: expected reference `&Rc<_>`
               found reference `&Arc<rustc_middle::traits::ObligationCauseCode<'_>>`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_trait_selection` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

estebank committed:
fix `clone` call

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3534170255) [(plain)](https://github.com/rust-lang/rust/commit/4fda83c69e1463ed8ea2479253b8fc70ee600a83/checks/3534170255/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
diff of stderr:

2   --> $DIR/issue-86530.rs:16:7
3    |
4 LL |     z(" ");
-    |       ^^^ the trait `X` is not implemented for `&str`
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
+    |     - ^^^ the trait `X` is not implemented for `&str`
+    |     required by a bound introduced by this call
6    |
7 note: required by a bound in `z`
8   --> $DIR/issue-86530.rs:10:8
---
To only update this specific test, also pass `--test-args const-generics/issues/issue-86530.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/issues/issue-86530.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/issues/issue-86530" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/issues/issue-86530/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0277]: the trait bound `&str: X` is not satisfied
   |
   |
LL |     z(" ");
   |     - ^^^ the trait `X` is not implemented for `&str`
   |     required by a bound introduced by this call
   |
note: required by a bound in `z`
  --> /checkout/src/test/ui/const-generics/issues/issue-86530.rs:10:8
  --> /checkout/src/test/ui/const-generics/issues/issue-86530.rs:10:8
   |
LL | fn z<T>(t: T)
   |    - required by a bound in this
LL | where
LL |     T: X,
   |        ^ required by this bound in `z`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.

---
test result: FAILED. 12008 passed; 1 failed; 102 ignored; 0 measured; 0 filtered out; finished in 131.27s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:28
```

</details>

estebank committed:
fix rebase

estebank committed:
Remove unnecessary label

bors commented:
:umbrella: The latest upstream changes (presumably #80522) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank committed:
Fix rebase

estebank commented on src/test/ui/suggestions/slice-issue-87994.stderr:
```diff
@@ -2,12 +2,9 @@ error[E0277]: the size for values of type `[i32]` cannot be known at compilation
-   |            help: consider borrowing here: `&v[1..]`
+   |            ^^^^^^ doesn't have a size known at compile-time
    |
-   = note: the trait bound `[i32]: IntoIterator` is not satisfied
+   = help: the trait `Sized` is not implemented for `[i32]`
```
This is a regression that will need to be fixed before landing.

nagisa reviewed (commented):
Are there any cases where we introduce the "required by a bound introduced by this call" note on something else than the immediate method being called? So far all of the test cases seem to annotate the method and the source of the obligation seems pretty obvious. Due to this note the presentation of certain diagnostics such as the ones involving closures become fairly awkward due to the overlapping spans.

Seems good from the implementation perspective. Definitely looks cleaner than passing an argument around.

estebank commented:
> Are there any cases where we introduce the "required by a bound introduced by this call" note on something else than the immediate method being called? So far all of the test cases seem to annotate the method and the source of the obligation seems pretty obvious. Due to this note the presentation of certain diagnostics such as the ones involving closures become fairly awkward due to the overlapping spans.

I wanted to make it clear that the reason the argument had some bound imposed on it was because of the call. This might not be needed given that have the `span_note` immediately afterwards in (I think) all cases, so I can remove it (or leave it with a span label with no text).

If by the overlapping case you are referring to is https://github.com/rust-lang/rust/pull/88719/files/ec904af0fc3d1dcbbd3514139e7cf6f31ce0c58b#diff-003f5fb5b17c96d98b8183c313df38486269166791f27003c3b62a4bfe251b81R8-R20, then I would say that that error is going to look ugly no matter what :-/ If we don't point to the call, that error will be _less_ clear, IMO. If you were referring to cases like https://github.com/rust-lang/rust/pull/88719/files/ec904af0fc3d1dcbbd3514139e7cf6f31ce0c58b#diff-a6816ee609a762d3ac2a91869bb43e3d53ed29359a9259b011664dc961dcd52a, then I agree, and am somewhat surprised at the regression: the `expected` span should point at `annotate` only, not the whole expression, but thought that could be dealt with later.

estebank commented on src/test/ui/associated-types/associated-types-path-2.stderr:
```diff
@@ -25,13 +27,15 @@ error[E0277]: the trait bound `u32: Foo` is not satisfied
    |
 LL |     f1(2u32, 4u32);
-   |     ^^^^^^^^^^^^^^ the trait `Foo` is not implemented for `u32`
+   |     ^^ the trait `Foo` is not implemented for `u32`
```
This is already what we had. Because of the new check, this error and the previous one got swapped. We now complain about the associated type first on the second argument, and then we complain about the materialized type itself when we evaluate the first argument. `pub fn f1<T: Foo>(a: T, x: T::A) {}` I think that because both arguments reference `T` in some way, we point at the call (as we did before this PR). I think that's reasonable, if not perfect, behavior.

estebank replied:
Interesting, looking at this closely, _this_ case doesn't point at the trait bound originating the obligation... It's likely that we don't have the right `ObligationCauseCode` for it.

estebank replied:
Resolved in the last commit

nagisa replied:
Very nice!

estebank committed:
Point at argument when evaluating `Path`'s bounds

When evaluating an `ExprKind::Call`, we first have to `check_expr` on it's
callee. When this one is a `ExprKind::Path`, we had to evaluate the bounds
introduced for its arguments, but by the time we evaluated them we no
longer had access to the argument spans. Now we special case this so
that we can point at the right place on unsatisfied bounds. This also
allows the E0277 deduplication to kick in correctly, so we now emit
fewer errors.

estebank commented on src/test/ui/where-clauses/where-clauses-method-unsatisfied.stderr:
```diff
@@ -2,7 +2,9 @@ error[E0277]: the trait bound `Bar: Eq` is not satisfied
    |
 LL |     x.equals(&x);
-   |              ^^ the trait `Eq` is not implemented for `Bar`
+   |       ------ ^^ the trait `Eq` is not implemented for `Bar`
+   |       |
+   |       required by a bound introduced by this call
```
This case has a `MiscObligation`, so for some `where` clauses we need to do better, but at least we point at the relevant argument. The new span pointing at the call at least hints at the user that they should look at the `equals` implementation to figure out why the bound was added.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 7a8de2eca8f9b2707ec9a6ac25fcfdfc845a1227 has been approved by `nagisa`

<!-- @bors r=nagisa 7a8de2eca8f9b2707ec9a6ac25fcfdfc845a1227 -->
<!-- homu: {"type":"Approved","sha":"7a8de2eca8f9b2707ec9a6ac25fcfdfc845a1227","approver":"nagisa"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/82183#issuecomment-920157843) for pull requests below priority 100. This pull request will be tested once the tree is reopened.

nagisa commented:
@bors rollup=never

bors commented:
:hourglass: Testing commit 7a8de2eca8f9b2707ec9a6ac25fcfdfc845a1227 with merge 633c54588419b027ba35ce87ff6ba15061de8ad7...
<!-- homu: {"type":"BuildStarted","head_sha":"7a8de2eca8f9b2707ec9a6ac25fcfdfc845a1227","merge_sha":"633c54588419b027ba35ce87ff6ba15061de8ad7"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3619361641)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3619361641","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3619207420) [(plain)](https://github.com/rust-lang-ci/rust/commit/633c54588419b027ba35ce87ff6ba15061de8ad7/checks/3619207420/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] ui/rfc-2632-const-trait-impl/const-drop-fail.rs#precise stdout ----
diff of stderr:

9 error[E0277]: the trait bound `NonTrivialDrop: Drop` is not satisfied
11    |
11    |
+ LL |         const _: () = check($exp);
+    |                       ----- required by a bound introduced by this call
+ ...
12 LL |     NonTrivialDrop,
13    |     ^^^^^^^^^^^^^^ the trait `Drop` is not implemented for `NonTrivialDrop`


21 error[E0277]: the trait bound `ConstImplWithDropGlue: Drop` is not satisfied
23    |
23    |
+ LL |         const _: () = check($exp);
+    |                       ----- required by a bound introduced by this call
+ ...
24 LL |     ConstImplWithDropGlue(NonTrivialDrop),
25    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Drop` is not implemented for `ConstImplWithDropGlue`


45 error[E0277]: the trait bound `NonTrivialDrop: A` is not satisfied
47    |
47    |
+ LL |         const _: () = check($exp);
+    |                       ----- required by a bound introduced by this call
+ ...
48 LL |     ConstDropImplWithBounds::<NonTrivialDrop>(PhantomData),
49    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `A` is not implemented for `NonTrivialDrop`


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2632-const-trait-impl/const-drop-fail.precise/const-drop-fail.precise.stderr
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
To only update this specific test, also pass `--test-args rfc-2632-const-trait-impl/const-drop-fail.rs`


error in revision `precise`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/rfc-2632-const-trait-impl/const-drop-fail.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "precise" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2632-const-trait-impl/const-drop-fail.precise" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2632-const-trait-impl/const-drop-fail.precise/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: `~const` is not allowed here
   |
   |
LL | struct ConstDropImplWithBounds<T: ~const A>(PhantomData<T>);
   |
   |
   = note: only allowed on bounds on traits' associated types and functions, const fns, const impls and its associated functions

error[E0277]: the trait bound `NonTrivialDrop: Drop` is not satisfied
   |
   |
LL |         const _: () = check($exp);
   |                       ----- required by a bound introduced by this call
...
LL |     NonTrivialDrop,
   |     ^^^^^^^^^^^^^^ the trait `Drop` is not implemented for `NonTrivialDrop`
note: required by a bound in `check`
  --> /checkout/src/test/ui/rfc-2632-const-trait-impl/const-drop-fail.rs:36:19
   |
   |
LL | const fn check<T: ~const Drop>(_: T) {}
   |                   ^^^^^^^^^^^ required by this bound in `check`

error[E0277]: the trait bound `ConstImplWithDropGlue: Drop` is not satisfied
   |
   |
LL |         const _: () = check($exp);
   |                       ----- required by a bound introduced by this call
...
LL |     ConstImplWithDropGlue(NonTrivialDrop),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Drop` is not implemented for `ConstImplWithDropGlue`
note: required by a bound in `check`
  --> /checkout/src/test/ui/rfc-2632-const-trait-impl/const-drop-fail.rs:36:19
   |
   |
LL | const fn check<T: ~const Drop>(_: T) {}
   |                   ^^^^^^^^^^^ required by this bound in `check`

error[E0277]: the trait bound `NonTrivialDrop: A` is not satisfied
   |
   |
LL |     ConstDropImplWithBounds::<NonTrivialDrop>(PhantomData),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `A` is not implemented for `NonTrivialDrop`
   |
note: required by `ConstDropImplWithBounds`
   |
   |
LL | struct ConstDropImplWithBounds<T: ~const A>(PhantomData<T>);


error[E0277]: the trait bound `NonTrivialDrop: A` is not satisfied
   |
   |
LL |         const _: () = check($exp);
   |                       ----- required by a bound introduced by this call
...
LL |     ConstDropImplWithBounds::<NonTrivialDrop>(PhantomData),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `A` is not implemented for `NonTrivialDrop`
   |
note: required by a bound in `ConstDropImplWithBounds`
   |
   |
LL | struct ConstDropImplWithBounds<T: ~const A>(PhantomData<T>);
   |                                   ^^^^^^^^ required by this bound in `ConstDropImplWithBounds`
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0277`.


------------------------------------------


---- [ui] ui/rfc-2632-const-trait-impl/const-drop-fail.rs#stock stdout ----
diff of stderr:

9 error[E0277]: the trait bound `NonTrivialDrop: Drop` is not satisfied
11    |
11    |
+ LL |         const _: () = check($exp);
+    |                       ----- required by a bound introduced by this call
+ ...
12 LL |     NonTrivialDrop,
13    |     ^^^^^^^^^^^^^^ the trait `Drop` is not implemented for `NonTrivialDrop`


21 error[E0277]: the trait bound `ConstImplWithDropGlue: Drop` is not satisfied
23    |
23    |
+ LL |         const _: () = check($exp);
+    |                       ----- required by a bound introduced by this call
+ ...
24 LL |     ConstImplWithDropGlue(NonTrivialDrop),
25    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Drop` is not implemented for `ConstImplWithDropGlue`


45 error[E0277]: the trait bound `NonTrivialDrop: A` is not satisfied
47    |
47    |
+ LL |         const _: () = check($exp);
+    |                       ----- required by a bound introduced by this call
+ ...
48 LL |     ConstDropImplWithBounds::<NonTrivialDrop>(PhantomData),
49    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `A` is not implemented for `NonTrivialDrop`


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2632-const-trait-impl/const-drop-fail.stock/const-drop-fail.stock.stderr
To only update this specific test, also pass `--test-args rfc-2632-const-trait-impl/const-drop-fail.rs`


error in revision `stock`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/rfc-2632-const-trait-impl/const-drop-fail.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "stock" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2632-const-trait-impl/const-drop-fail.stock" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2632-const-trait-impl/const-drop-fail.stock/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: `~const` is not allowed here
   |
   |
LL | struct ConstDropImplWithBounds<T: ~const A>(PhantomData<T>);
   |
   |
   = note: only allowed on bounds on traits' associated types and functions, const fns, const impls and its associated functions

error[E0277]: the trait bound `NonTrivialDrop: Drop` is not satisfied
   |
   |
LL |         const _: () = check($exp);
   |                       ----- required by a bound introduced by this call
...
LL |     NonTrivialDrop,
   |     ^^^^^^^^^^^^^^ the trait `Drop` is not implemented for `NonTrivialDrop`
note: required by a bound in `check`
  --> /checkout/src/test/ui/rfc-2632-const-trait-impl/const-drop-fail.rs:36:19
   |
   |
LL | const fn check<T: ~const Drop>(_: T) {}
   |                   ^^^^^^^^^^^ required by this bound in `check`

error[E0277]: the trait bound `ConstImplWithDropGlue: Drop` is not satisfied
   |
   |
LL |         const _: () = check($exp);
   |                       ----- required by a bound introduced by this call
...
LL |     ConstImplWithDropGlue(NonTrivialDrop),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Drop` is not implemented for `ConstImplWithDropGlue`
note: required by a bound in `check`
  --> /checkout/src/test/ui/rfc-2632-const-trait-impl/const-drop-fail.rs:36:19
   |
   |
LL | const fn check<T: ~const Drop>(_: T) {}
   |                   ^^^^^^^^^^^ required by this bound in `check`

error[E0277]: the trait bound `NonTrivialDrop: A` is not satisfied
   |
   |
LL |     ConstDropImplWithBounds::<NonTrivialDrop>(PhantomData),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `A` is not implemented for `NonTrivialDrop`
   |
note: required by `ConstDropImplWithBounds`
   |
   |
LL | struct ConstDropImplWithBounds<T: ~const A>(PhantomData<T>);


error[E0277]: the trait bound `NonTrivialDrop: A` is not satisfied
   |
   |
LL |         const _: () = check($exp);
   |                       ----- required by a bound introduced by this call
...
LL |     ConstDropImplWithBounds::<NonTrivialDrop>(PhantomData),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `A` is not implemented for `NonTrivialDrop`
   |
note: required by a bound in `ConstDropImplWithBounds`
   |
   |
LL | struct ConstDropImplWithBounds<T: ~const A>(PhantomData<T>);
   |                                   ^^^^^^^^ required by this bound in `ConstDropImplWithBounds`
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0277`.

---
test result: FAILED. 12036 passed; 2 failed; 102 ignored; 0 measured; 0 filtered out; finished in 129.75s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:21
```

</details>

estebank committed:
fix rebase

estebank commented:
@bors r=nagisa

bors commented:
:pushpin: Commit 0a4540b08f4d43c8d8d9f4de3d78daa5fcfd8b8f has been approved by `nagisa`

<!-- @bors r=nagisa 0a4540b08f4d43c8d8d9f4de3d78daa5fcfd8b8f -->
<!-- homu: {"type":"Approved","sha":"0a4540b08f4d43c8d8d9f4de3d78daa5fcfd8b8f","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 0a4540b08f4d43c8d8d9f4de3d78daa5fcfd8b8f with merge e36621057d9f497c822eb800934b5933c10510cf...
<!-- homu: {"type":"BuildStarted","head_sha":"0a4540b08f4d43c8d8d9f4de3d78daa5fcfd8b8f","merge_sha":"e36621057d9f497c822eb800934b5933c10510cf"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3627539321)
Approved by: nagisa
Pushing e36621057d9f497c822eb800934b5933c10510cf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3627539321"},"merge_sha":"e36621057d9f497c822eb800934b5933c10510cf"} -->

rust-timer commented:
Finished benchmarking commit (e36621057d9f497c822eb800934b5933c10510cf): [comparison url](https://perf.rust-lang.org/compare.html?start=e4828d5b7f745a9e867a9b0cc7f080f287bcf55d&end=e36621057d9f497c822eb800934b5933c10510cf).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

 

@rustbot label: -perf-regression

pnkfelix commented:
Huh. I'm doing performance triage now, and the triage report points to this PR as the source of a [1.9% regression on diesel](https://perf.rust-lang.org/compare.html?start=e4828d5b7f745a9e867a9b0cc7f080f287bcf55d&end=e36621057d9f497c822eb800934b5933c10510cf).

There has been some associated discussion of the matter on zulip : https://zulip-archive.rust-lang.org/stream/247081-t-compiler/performance/topic/triage.202021.2021.2009.html#254207973

Anyway, it seems like this might be a legitimate regression to investigate, at least for the diesel-doc case.

@rustbot label: +perf-regression
